pkg/webhooks/webhookpatch.go

```diff
@@ -17,133 +17,164 @@ package webhooks
 			},
 		},
 	)
-	go controller.Run(stopCh)
 
-	go func() {
-		var delayedRetryC <-chan time.Time
-		if retry {
-			delayedRetryC = time.After(delayedRetryTime)
-		}
+	c.Run(stopChan)
+}
 
-		for {
-			select {
-			case <-delayedRetryC:
-				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
-					delayedRetryC = time.After(delayedRetryTime)
-				} else {
-					log.Infof("Retried patch succeeded")
-					delayedRetryC = nil
-				}
-			case <-shouldPatch:
-				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
-					if delayedRetryC == nil {
-						delayedRetryC = time.After(delayedRetryTime)
-					}
-				} else {
-					delayedRetryC = nil
-				}
+func (w *WebhookCertPatcher) updateWebhookHandler(oldConfig, newConfig *v1beta1.MutatingWebhookConfiguration) {
+	if oldConfig.ResourceVersion != newConfig.ResourceVersion {
+		for i, wh := range newConfig.Webhooks {
+			if wh.Name == w.webhookName && !bytes.Equal(newConfig.Webhooks[i].ClientConfig.CABundle, w.caCertPem) {
+				w.queue.Push(func() error {
+					return w.webhookPatchTask(newConfig.Name)
+				})
+				break
 			}
 		}
-	}()
+	}
 }
 
-func doPatch(cs kubernetes.Interface, webhookConfigName, webhookName string, caCertPem []byte) (retry bool) {
-	client := cs.AdmissionregistrationV1beta1().MutatingWebhookConfigurations()
-	if err := patchMutatingWebhookConfig(client, webhookConfigName, webhookName, caCertPem); err != nil {
-		log.Errorf("Patch webhook failed: %v", err)
-		return true
+func (w *WebhookCertPatcher) addWebhookHandler(config *v1beta1.MutatingWebhookConfiguration) {
+	for i, wh := range config.Webhooks {
+		if wh.Name == w.webhookName && !bytes.Equal(config.Webhooks[i].ClientConfig.CABundle, w.caCertPem) {
+			log.Infof("New webhook config added, patching MutatingWebhookConfiguration for %s", config.Name)
+			w.queue.Push(func() error {
+				return w.webhookPatchTask(config.Name)
+			})
+			break
+		}
+	}
+}
+
+// webhookPatchTask takes the result of patchMutatingWebhookConfig and modifies the result for use in task queue
+func (w *WebhookCertPatcher) webhookPatchTask(webhookConfigName string) error {
+	err := w.patchMutatingWebhookConfig(
+		w.client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations(),
+		webhookConfigName)
+
+	// do not want to retry the task if these errors occur, they indicate that
+	// we should no longer be patching the given webhook
+	if errors.Is(err, errWrongRevision) || errors.Is(err, errNoWebhookWithName) {
+		return nil
+	}
+
+	return err
+}
+
+// patchMutatingWebhookConfig takes a webhookConfigName and patches the CA bundle for that webhook configuration
+func (w *WebhookCertPatcher) patchMutatingWebhookConfig(
+	client admissionregistrationv1beta1client.MutatingWebhookConfigurationInterface,
+	webhookConfigName string) error {
+	config, err := client.Get(context.TODO(), webhookConfigName, metav1.GetOptions{})
+	if err != nil {
+		return err
```

Comment:
if the webhook is deleted, I think this may loop? might need an `errors.IsNotFound` check?